### PR TITLE
Provide implementations for entity SDK used by orchestrators

### PR DESCRIPTION
**Pull request by sebastianburckhardt**
Provides the necessary shim classes to implement entity support for orchestration contexts.

### CODE REVIEW

File: src/Worker/Core/Shims/TaskOrchestrationShim.cs [hunk 1]
```diff
@@ -46,10 +46,19 @@ partial class TaskOrchestrationShim : TaskOrchestration
...
+        {
+            object? output = await this.implementation.RunAsync(this.wrapperContext, input);
+
+            // Return the output (if any) as a serialized string.
+            return this.DataConverter.Serialize(output);
+        }
+        finally
+        {
+            // if user code crashed inside a critical section, or did not exit it, do that now
+            this.wrapperContext.ExitCriticalSectionIfNeeded();
```
Comment (jviau) [inline src/Worker/Core/Shims/TaskOrchestrationShim.cs:60]: This works _for now_. But not ideal that this logic only happens as part of the wrapper shim. Fine for now I suppose.

File: src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+/// <summary>
+/// A wrapper to go from <see cref="OrchestrationEntityContext" /> to <see cref="TaskOrchestrationEntityFeature "/>.
+/// </summary>
+sealed class TaskOrchestrationEntityContext : TaskOrchestrationEntityFeature
+{
+    static readonly List<EntityInstanceId> EmptyEntityList = new List<EntityInstanceId>();
+
+    // we don't know the actual max delay, because we don't have backend information, so we choose
+    // a value conservatively that should work for all backends
+    static readonly TimeSpan MaxSignalDelay = TimeSpan.FromDays(6);
```
[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs:26]: Great, I am using that now.

File: src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    readonly TaskOrchestrationContextWrapper taskOrchestrationContext;
+    readonly OrchestrationContext innerContext;
+    readonly string instanceId;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TaskOrchestrationEntityContext"/> class.
+    /// </summary>
+    /// <param name="taskOrchestrationContext">The wrapper for the orchestration context.</param>
+    /// <param name="innerContext">The inner orchestration entity context.</param>
+    public TaskOrchestrationEntityContext(TaskOrchestrationContextWrapper taskOrchestrationContext, OrchestrationContext innerContext)
```
[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs:37]: Good idea!

File: src/Worker/Core/Shims/TaskOrchestrationShim.cs (hunk 1 above)
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationShim.cs:60]: Yeah, note that there are already some paths on which the lock fails to release (e.g. under nondeterministic orchestration errors). It is not completely avoidable unfortunately.

That's why we have the 'release orphaned locks' as a backup "solution".

File: src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        if (dtEntities.Length == 0)
+        {
+            throw new ArgumentException("The list of entities to lock must not be empty.", nameof(entityIds));
+        }
+
+        // use a deterministically replayable unique ID for this lock request, and to receive the response
+        var criticalSectionId = this.taskOrchestrationContext.NewGuid();
+
+        // send a message to the first entity to be acquired
+        var entityMessageEvent = this.EntityContext.EmitAcquireMessage(criticalSectionId, dtEntities);
```
[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs]: yep, going through these now. VS actually shows this to me in messages, I just noticed.

File: src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    }
+
+    static TaskFailureDetails ConvertFailureDetails(FailureDetails failureDetails)
+     => new TaskFailureDetails(
+         failureDetails.ErrorType,
+         failureDetails.ErrorMessage,
+         failureDetails.StackTrace,
+         failureDetails.InnerFailure != null ? ConvertFailureDetails(failureDetails.InnerFailure) : null);
+
+    async ValueTask<OperationResult> SignalOrCallEntityInternalAsync(EntityInstanceId id, string operationName, object? input, bool oneWay = false, DateTimeOffset? scheduledTime = null)
```
[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs]: Yes. This is only async when `oneWay=false`.

[the remark this answers is not in the document]
Comment (sebastianburckhardt) [reply src/Worker/Core/Shims/TaskOrchestrationEntityContext.cs]: Right. I will separate them.

**Review by jviau — APPROVED** [review]

**Pull request merged**